File: utils/rf_analysis.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.fft import fft, ifft

def fft_fun(signal, fs):
    # 进行傅里叶变换
    fft_result = np.fft.fft(signal)
    n = len(signal)
    freq = np.fft.fftfreq(n, d=1/fs)

    # 只取正频率部分
    indices = np.where(freq >= 0)
    freq = freq[indices]
    fft_magnitude = np.abs(fft_result[indices])

    fft_magnitude = fft_magnitude / np.max(fft_magnitude)

    return freq, fft_magnitude

def bandpass_filter_rf_data(x, length, sampling_frequency, low_cutoff, high_cutoff):
    size = int(length)

    fft_result = fft(x)

    df = sampling_frequency / size  # 计算频率分辨率

    # 创建频率数组
    frequencies = np.arange(size) * df

    # 应用带通滤波器
    mask = (frequencies >= low_cutoff) & (frequencies <= high_cutoff)
    fft_result[~mask] = 0.0  # 将不在范围内的频率分量置零

    res = ifft(fft_result)

    return np.real(res)



df = pd.read_csv(r'rfdata\rfdata_1_32.csv', sep=',', header=None)
data = df.values
data = (data - 512) / 512
data = data.T

# for i in range(data.shape[0]):
#     data[i, :] = bandpass_filter_rf_data(data[i, :], data.shape[1], 25e6, 1.0e6, 5.0e6)

fs = 25e6
# signal1 = data[32, 100:228]
# signal2 = data[32, 1000:1128]
# signal3 = data[32, 2000:2128]

# freq1, fft_magnitude1 = fft_fun(signal1, fs)
# freq2, fft_magnitude2 = fft_fun(signal2, fs)
# freq3, fft_magnitude3 = fft_fun(signal3, fs)

# for i in range(5):
#     di = i * 790
#     signal = data[32, di:di+128]
#     signal, fft_magnitude = fft_fun(signal, fs)
#     plt.plot(signal, fft_magnitude)

freq, fft_magnitude = fft_fun(data[33, :], fs)


# 绘制频谱图
# plt.figure(figsize=(10, 4))
plt.plot(freq, fft_magnitude)
# plt.plot(freq2, fft_magnitude2)
# plt.plot(freq3, fft_magnitude3)
plt.xlabel('Frequency (Hz)')
plt.ylabel('Amplitude')
plt.xlim(0, 8e6)
plt.grid(True)
plt.show()

File: utils/rf_analysis.m
% 1. 读取 CSV 文件数据（推荐使用 readmatrix，适用于纯数值文件）
data = readmatrix('rfdata\rfdata_1_32.csv');
data = (data - 512) / 512;
data = bandpass_filter(data);

x = data(3000:3500, 32);
fs = 20e6;
L = length(x);
T = 1/ fs;
t = (0 : L-1)*T;

Y = fft(x);

Y = abs(Y/L);
P1 = Y(1 : L/2+1);
P1(2:end-1) = 2*P1(2:end-1);

f = fs * (0 : (L/2)) / L;

% 绘制频谱图
figure;
plot(f/1e6, P1);    % 将频率转换为MHz单位
title('单边幅值谱');
xlabel('频率 (MHz)');
ylabel('|幅值|');
grid on;

% 显示主要频率峰值
[~, idx1] = max(P1);
f_peak1 = f(idx1)/1e6;
disp(['检测到峰值频率：', num2str(f_peak1), ' MHz']);


% =================================================================
function filtered_signal = bandpass_filter(signal)
    % 带通滤波器函数
    % 输入参数：
    %   signal - 待滤波的信号
    % 输出参数：
    %   filtered_signal - 滤波后的信号

    % 设置采样频率
    fs = 20e6;  % 20 MHz

    % 设置带通滤波器的截止频率（1MHz 到 5MHz）
    low_cutoff = 0.5e6;
    high_cutoff = 5e6;

    % 归一化截止频率（归一化至 Nyquist 频率 fs/2）
    Wn = [low_cutoff high_cutoff] / (fs/2);

    % 选择滤波器阶数（可根据具体需求调整）
    order = 4;

    % 使用 Butterworth 方法设计带通滤波器
    [b, a] = butter(order, Wn, 'bandpass');

    % 使用 filtfilt 进行零相位滤波，避免相位延迟
    filtered_signal = filtfilt(b, a, signal);
end
